Added CaretBlinkInterval to TextPresenter

sLill opened the pull request:
<!--- See CONTRIBUTING.md for general guidelines on contributions -->

## What does the pull request do?
<!--- Give a bit of background on the PR here, together with links to with related issues etc. -->
Replaces hard-coded caret  blink interval in TextPresenter with CaretBlinkInterval StyledProperty

## What is the current behavior?
<!--- If the PR is a fix, describe the current incorrect behavior, otherwise delete this section. -->
TextPresenter caret blinks at a constant hard-coded rate of 500ms 

## What is the updated/expected behavior with this PR?
<!--- Describe how to test the PR. -->
Tested in sample projects found in \Avalonia\samples\

## How was the solution implemented (if it's not obvious)?
<!--- Include any information that might be of use to a reviewer here. -->
CaretBlinkInterval StyledProperty is given a default value of 500ms to keep its' behavior consistent with how it worked before. Setting the interval <= 0 disables the blink altogether

## Checklist

- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/avalonia-docs with user documentation

## Breaking changes
<!--- List any breaking changes here. -->

## Obsoletions / Deprecations
<!--- Obsolete and Deprecated attributes on APIs MUST only be included when discussed with Core team. @grokys, @kekekeks & @danwalmsley -->

## Fixed issues
<!--- If the pull request fixes issue(s) list them like this: 
Fixes #123
Fixes #456
-->
Fixes https://github.com/AvaloniaUI/Avalonia/issues/13681

sLill committed:
Added CaretBlinkInterval to TextPresenter

sLill committed:
TextPresenter CaretBlinkInterval property exposed through TextBox

sLill commented:
> @sLill please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @dotnet-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @dotnet-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @dotnet-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@dotnet-policy-service agree

Gillibald commented:
What happens if someone changes the interval? You are not handling changes at the moment. Pls add that case to the PropertyChanged handler

maxkatz6 commented:
It would be nice to have TimeSpan property. XAML compiler already handles compile-time parsing of time spans as well.

sLill committed:
Added PropertyChanged handler for CaretBlinkInterval

sLill committed:
CaretBlinkInterval datatype changed from int to TimeSpan

sLill commented:
> What happens if someone changes the interval? You are not handling changes at the moment. Pls add that case to the PropertyChanged handler

Fixed

sLill commented:
> It would be nice to have TimeSpan property. XAML compiler already handles compile-time parsing of time spans as well.

Fixed. Changed datatype from int to TimeSpan

robloo commented on src/Avalonia.Controls/TextBox.cs:
```diff
@@ -443,6 +449,15 @@ public char PasswordChar
         }
 
+        /// <summary>
+        /// Gets or sets a brush that is used for the text caret
```
Wrong comments (you copy/pasted), missing period.

sLill replied:
Fixed

robloo commented on src/Avalonia.Controls/Presenters/TextPresenter.cs:
```diff
@@ -34,6 +34,9 @@ public class TextPresenter : Control
         public static readonly StyledProperty<IBrush?> CaretBrushProperty =
             AvaloniaProperty.Register<TextPresenter, IBrush?>(nameof(CaretBrush));
 
+        public static readonly StyledProperty<TimeSpan> CaretBlinkIntervalProperty =
+            AvaloniaProperty.Register<TextPresenter, TimeSpan>(nameof(CaretBlinkInterval), defaultValue: TimeSpan.FromMilliseconds(500));
+
```
@Gillibald @maxkatz6 What is going on with Caret-related properties between TextPresenter and TextBox?

In a perfect world they should be factored out into a base class (which has been discussed elsewhere). Here though some properties are addowner from TextBox or vice-versa. TextPresenter seems lower level so actually TextBox should AddOwner this property. Overall, this really needs a thorough review and possibly noted for a v12 breaking change cleanup.

Then for this PR we need to decide if TextPresenter or TextBox should have the property and then AddOwner accordingly.

sLill replied:
@robloo @Gillibald @maxkatz6 If you all would like to do a major refactor of caret properties, that's likely out of the scope of this ticket. So you are welcome to close this PR and address it in a broader scope or merge this in the meantime- no hard feelings either way.

robloo replied:
No, as mentioned earliest would be v12 as it's a breaking change -- well after this PR. The property inheritance here (or lack thereof) doesn't make sense to me but need others feedback to confirm.

sLill replied:
Ah I see. I agree with that observation then. That area could use some work in the near future.

Gillibald replied:
TextPresenter is something I want to remove entirely. Syncing state via bindings is a total nightmare.

robloo commented on src/Avalonia.Controls/Presenters/TextPresenter.cs:
```diff
@@ -288,6 +288,12 @@ public bool RevealPassword
             set => SetValue(CaretBrushProperty, value);
         }
 
+        public TimeSpan CaretBlinkInterval
```
Missing comments which should always be added for new properties. They can be inherited from TextBox (or more likely TextBox inherits from here).

sLill replied:
I'm familiar with using <inheritdoc/> to inherit comments from base classes- but since TextBox just uses a TextPresenter and doesn't necessarily inherit from it, I'm not sure how this can be done. For now, I've copied the comment over.

jp2masa replied:
Something like this should work: https://github.com/AvaloniaUI/Avalonia/pull/13567#discussion_r1389514933.

sLill replied:
> Something like this should work: [#13567 (comment)](https://github.com/AvaloniaUI/Avalonia/pull/13567#discussion_r1389514933).

Did not know this was possible. Fixed it. Thank you.

sLill committed:
Added CaretBlinkInterval comments

sLill committed:
TextBox.CaretBlinkInterval comment changed to inherit from TextPresenter.CaretBlinkInterval

Gillibald committed:
Merge branch 'master' into caret_blink_interval

Gillibald commented on src/Avalonia.Controls/Presenters/TextPresenter.cs:
```diff
@@ -34,6 +34,9 @@ public class TextPresenter : Control
             AvaloniaProperty.Register<TextPresenter, IBrush?>(nameof(CaretBrush));
 
+        public static readonly StyledProperty<TimeSpan> CaretBlinkIntervalProperty =
+            AvaloniaProperty.Register<TextPresenter, TimeSpan>(nameof(CaretBlinkInterval), defaultValue: TimeSpan.FromMilliseconds(500));
```
You can use AddOwner here so the property is only defined at one location

sLill replied:
Fixed. Uses AddOwner now.

sLill committed:
AddOwner given to CaretBlinkInterval in TextPresenter
